### ISSUE TITLE
Compile compat issue of ppc at linux_debug.c, added: reg/linux_ppc.h tested with: configure --disable-debugger at sys/build.sh

### ISSUE THREAD

**Issue by unixfreaxjp** [radare/radare2#4720]
On cpu:
```asm
cpu             : 740/750
temperature     : 62-64 C (uncalibrated)
clock           : 266.000000MHz
revision        : 3.1 (pvr 0008 0301)
bogomips        : 33.20
timebase        : 16601533
platform        : PowerMac
model           : Power Macintosh
machine         : Power Macintosh
motherboard     : AAPL,PowerMac G3 MacRISC
detected as     : 49 (PowerMac G3 (Silk))
pmac-generation : OldWorld
Memory          : 256 MB
```
Compilation was tested using several methods & permission:
```
1. sys/user.sh (non root)
2. sys/install.sh (root)
3. cd ~/radare2&&./configure&&make
```
After several "minor" *warnings*, the error is getting more serious..
This is where *serious* errors came up..."e.g. unsupported cpu??"
```asm
[...]
p/native/linux/linux_debug.c: In function linux_reg_profile・
p/native/linux/linux_debug.c:28:2: error: #error "Unsupported Linux CPU"
p/native/linux/linux_debug.c: In function linux_handle_signals・
p/native/linux/linux_debug.c:43:2: warning: #warning DO MORE RDEBUGREASON HERE [-Wcpp]
p/native/linux/linux_debug.c: In function print_fpu・
p/native/linux/linux_debug.c:291:2: warning: #warning not implemented for this platform [-Wcpp]
p/native/linux/linux_debug.c: In function linux_reg_read・
p/native/linux/linux_debug.c:376:3: warning: #warning not implemented for this platform [-Wcpp]
p/native/linux/linux_debug.c:383:4: error: unknown type name _DEBUG_REG_T・
p/native/linux/linux_debug.c:296:7: warning: variable howfpu・set but not used [-Wunused-but-set-variable]
p/native/linux/linux_debug.c: In function linux_reg_write・
p/native/linux/linux_debug.c:425:43: error: struct user・has no member named 穹_debugreg・
p/native/linux/linux_debug.c:431:17: error: _DEBUG_REG_T・undeclared (first use in this function)
p/native/linux/linux_debug.c:431:17: note: each undeclared identifier is reported only once for each function it appears in
p/native/linux/linux_debug.c:444:48: error: egs・undeclared (first use in this function)
CC asm_malbolge.c
p/native/linux/linux_debug.c: At top level:
p/native/linux/linux_debug.c:203:13: warning: print_fpu・defined but not used [-Wunused-function]
CC asm_mcs96.c
p/native/linux/linux_debug.c: In function Linux_reg_profile・
p/native/linux/linux_debug.c:30:1: warning: control reaches end of non-void function [-Wreturn-type]
make[4]: *** [p/native/linux/linux_debug.o] Error 1
make[4]: *** Waiting for unfinished jobs....
CC asm_mips_cs.c
CC asm_mips_gnu.c
CC mips-dis.c
make[3]: *** [foo] Error 2
make[2]: *** [debug] Error 2
make[2]: *** Waiting for unfinished jobs....
CC mips16-opc.c
CC mips-opc.c
CC mipsasm.c
[...]
```
And this is the "kills" for the compilation session..
```asm
CC elf32-xtensa.c
CC asm_z80.c
CC asm_z80_cr.c
CC asm.c
CC code.c
asm.c: In function xxx_asm_new・
asm.c:131:14: warning: unused variable static_plugin・[-Wunused-variable]
LD libr_asm.so
CC asm_propeller.c
CC propeller_disas.c
gcc: warning: /test/radare2/libr/../shlr/capstone/libcapstone.a: linker input file unused because linking not done
gcc: warning: /test/radare2/libr/../shlr/capstone/libcapstone.a: linker input file unused because linking not done
make[1]: *** [all] Error 2
make: *** [all] Error 2
Oops
```
Suggested problem:
```
Compatibility problem? Unsupported cpu? Overlapping in linking before "make" process itself?
(please enlight us)
```
More longer compilation log is in here:
https://github.com/unixfreaxjp/dev/blob/master/r2ppc.md

**Comment by crowell** [comment]
the unsupported CPU message comes from here:

```c
const char *linux_reg_profile (RDebug *dbg) {
#if __arm__
#include "reg/linux-arm.h"
#elif __arm64__ || __aarch64__
#include "reg/linux-arm64.h"
#elif __MIPS__ || __mips__
#include "reg/linux-mips.h"
#elif (__i386__ || __x86_64__)
  if (dbg->bits & R_SYS_BITS_32) {
#if __x86_64__
#include "reg/linux-x64-32.h"
#else
#include "reg/linux-x86.h"
#endif
  } else {
#include "reg/linux-x64.h"
  }
#else
#error "Unsupported Linux CPU"
#endif
}
```
can you try this patch? https://gist.github.com/ae283ea135d5be56388042f49d24418c

**Comment by crowell** [comment]
(sparc should also fail because it doesn't have a reg profile in the debug plugin.)

**Comment by unixfreaxjp** [comment]
> can you try this patch? https://gist.github.com/ae283ea135d5be56388042f49d24418c

okay, trying this.

![](https://lh3.googleusercontent.com/-S3VpBidVApg/VyC3rhqZ6cI/AAAAAAAAUyg/Km23CGbxgwMtZo6FE3klFvOaSrWQyfPAgCLcB/s600/001.PNG)

And also trying this one too. Hang on gents, the cpu is slow. On it. Will report.

**Comment by unixfreaxjp** [comment]
The option --disable-debugger configured well:
```asm
Final report:
 - PREFIX = /test/bin/prefix/radare2/
 - HAVE_LIB_GMP = 0
 - HAVE_OPENSSL = 0
 - USE_CAPSTONE = 0
 - HAVE_FORK = 1
 - VERSION = 0.10.3-git
 - USE_LIB_ZIP = 0
 - USE_LIB_MAGIC = 0
 - DEBUGGER = 0 <=========================== Yes!
 - CC = gcc
 - USERCC = gcc
 - USEROSTYPE = gnulinux
 - LIL_ENDIAN = 0
 - LIBVERSION = 0.10.3-git
Generating r_version.h file [...]
```
Build environment is correct too:
```asm
BUILD SUMARY
============
COMPILER gcc
CC gcc
HOST_CC gcc
HOST_OS linux
BUILD_OS linux
============
But it can not built, somehow during the compilation process **it still linked to native Linux compiler **even the debugger is disabled. Below is the collected warning snipped until it stop the build due to error. 

```
[...]

calc.c: In function 'r_hash_calculate':
calc.c:39:8: warning: unused variable 'pres' [-Wunused-variable]

[...]

CC cr16_disas.c
/test/radare2/libr/..//libr/bin/p/bin_elf.c:581:15: warning: 'patch_relocs' defined but not used [-Wunused-function]
CC dex.c

[...]

[...]

CC utils.c
CC hashtable.c
CC hashvector.c
CC pe64.c
CC bin_pebble.c
CC bin_psxexe.c
/test/radare2/libr/..//libr/anal/p/../..//asm/arch/tms320/c55x_plus/hashvector.c:97:25: warning: 'ins_hash' defined but not used [-Wunused-variable]
CC anal_v810.c

[...]

CC demangler.c
CC microsoft_demangle.c
/test/radare2/libr/..//libr/anal/p/../../asm/arch/z80_cr/z80_tab.h:936:19: warning: 'z80_op' defined but not used [-Wunused-variable]
CC objc.c

[...]

CC anal_ex.c
CC switch.c
anal.c: In function 'r_anal_set_reg_profile':
anal.c:205:7: warning: variable 'ret' set but not used [-Wunused-but-set-variable]
CC state.c

[...]

CC mmemory.c
CC parse_ppc_pseudo.c
CC parse_x86_pseudo.c
LD libr_parse.so
CC parse_z80_pseudo.c
gcc: warning: /test/radare2/libr/../shlr/zip/librz.a: linker input file unused because linking not done
DIR asm
DIR debug
CC asm_6502.c
CC asm_8051.c

[...] // here comes the debugger... why it is compiled even it was configured as disabled?

CC debug_bf.c
CC asm_cris_gnu.c
CC cris-dis.c
CC debug_bochs.c
CC debug_esil.c
CC cris-opc.c
CC asm_csr.c
CC asm_dalvik.c
CC asm_dcpu16.c
CC debug_gdb.c
CC asm_ebc.c
CC debug_native.c
CC debug_qnx.c
CC asm_gb.c
CC asm_h8300.c
CC asm_hppa_gnu.c
CC debug_rap.c
CC hppa-dis.c
CC debug_wind.c
CC asm_i4004.c
CC linux_debug.c
CC asm_i8080.c
CC asm_java.c
CC asm_lanai_gnu.c
CC lanai-dis.c
CC lanai-opc.c
CC asm_lh5801.c
CC asm_lm32.c
p/native/linux/linux_debug.c: In function 'linux_reg_profile':
p/native/linux/linux_debug.c:28:2: error: #error "Unsupported Linux CPU"
p/native/linux/linux_debug.c: In function 'linux_handle_signals':
p/native/linux/linux_debug.c:43:2: warning: #warning DO MORE RDEBUGREASON HERE [-Wcpp]
p/native/linux/linux_debug.c: In function 'print_fpu':
p/native/linux/linux_debug.c:291:2: warning: #warning not implemented for this platform [-Wcpp]
p/native/linux/linux_debug.c: In function 'linux_reg_read':
p/native/linux/linux_debug.c:376:3: warning: #warning not implemented for this platform [-Wcpp]
p/native/linux/linux_debug.c:383:4: error: unknown type name 'R_DEBUG_REG_T'
p/native/linux/linux_debug.c:296:7: warning: variable 'showfpu' set but not used [-Wunused-but-set-variable]
p/native/linux/linux_debug.c: In function 'linux_reg_write':CC asm_m68k.c

p/native/linux/linux_debug.c:425:43: error: 'struct user' has no member named 'u_debugreg'
p/native/linux/linux_debug.c:431:17: error: 'R_DEBUG_REG_T' undeclared (first use in this function)
p/native/linux/linux_debug.c:431:17: note: each undeclared identifier is reported only once for each function it appears in
p/native/linux/linux_debug.c: At top level:
p/native/linux/linux_debug.c:203:13: warning: 'print_fpu' defined but not used [-Wunused-function]
p/native/linux/linux_debug.c: In function 'linux_reg_profile':
p/native/linux/linux_debug.c:30:1: warning: control reaches end of non-void function [-Wreturn-type]
make[4]: *** [p/native/linux/linux_debug.o] Error 1
make[4]: *** Waiting for unfinished jobs....
CC m68k_disasm.c
CC asm_m68k_cs.c
CC asm_malbolge.c
make[3]: *** [foo] Error 2
make[2]: *** [debug] Error 2
make[2]: *** Waiting for unfinished jobs....
CC asm_mcs96.c
CC asm_mips_cs.c
CC asm_mips_gnu.c
CC mips-dis.c
CC mips16-opc.c
CC mips-opc.c
CC mipsasm.c

[...]

CC asm_z80_cr.c
CC asm.c
CC code.c
LD libr_asm.so
CC asm_propeller.c
CC propeller_disas.c
gcc: warning: /home/mung/radare2/libr/../shlr/capstone/libcapstone.a: linker input file unused because linking not done
gcc: warning: /home/mung/radare2/libr/../shlr/capstone/libcapstone.a: linker input file unused because linking not done
make[1]: *** [all] Error 2
make: *** [all] Error 2
Oops

```
Result so far, the problem persisted, below advise is not fixing the problem, the error must be in other place.
```
1. This patch? https://gist.github.com/ae283ea135d5be56388042f49d24418c
2. ./configure --disable-debugger 
```
let's see the code slowly now.. really want to know why it still called native linux debugger..

**Comment by unixfreaxjp** [comment]
Is this the main problem?

https://lh3.googleusercontent.com/-Ovd-JgQLZhE/VyF7s2sqEDI/AAAAAAAAUzU/AnCaMzkVxIckh2akHqwQcg2l6zYIInMbQCLcB/s800/Capture.PNG

Re-checking the overall process, found that patch ae283ea135d5be56388042f49d24418c wasn't applied correctly..
Now re-applying patch to:

- libr/debug/p/native/linux/linux_debug.c

```cpp
[...]
#elif __ppc__
#include "reg/linux-ppc.h"
[...]
```
and added header file (for comparison xref: http://lxr.free-electrons.com/source/arch/powerpc/include/asm/reg.h)  

- libr/debug/p/native/linux/reg/linux-ppc.h

```cpp
return strdup (
"=PC	srr0\n"
"=SP	srr1\n"
"=SR	srr1\n" // status register ??
"=A0	r0\n"
"=A1	r1\n"
"=A2	r2\n"
"=A3	r3\n"
#if 0
"=A4	r4\n"
"=A5	r5\n"
"=A6	r6\n"
"=A7	r7\n"
#endif
"gpr	srr0	.32	0	0\n"
"gpr	srr1	.32	4	0\n"
"gpr	r0	.32	8	0\n"
"gpr	r1	.32	12	0\n"
"gpr	r2	.32	16	0\n"
"gpr	r3	.32	20	0\n"
"gpr	r4	.32	24	0\n"
"gpr	r5	.32	28	0\n"
"gpr	r6	.32	32	0\n"
"gpr	r7	.32	36	0\n"
"gpr	r8	.32	40	0\n"
"gpr	r9	.32	44	0\n"
"gpr	r10	.32	48	0\n"
"gpr	r11	.32	52	0\n"
"gpr	r12	.32	56	0\n"
"gpr	r13	.32	60	0\n"
"gpr	r14	.32	64	0\n"
"gpr	r15	.32	68	0\n"
"gpr	r16	.32	72	0\n"
"gpr	r17	.32	76	0\n"
"gpr	r18	.32	80	0\n"
"gpr	r19	.32	84	0\n"
"gpr	r20	.32	88	0\n"
"gpr	r21	.32	92	0\n"
"gpr	r22	.32	96	0\n"

"gpr	r23	.32	100	0\n"
"gpr	r24	.32	104	0\n"
"gpr	r25	.32	108	0\n"
"gpr	r26	.32	112	0\n"
"gpr	r27	.32	116	0\n"
"gpr	r28	.32	120	0\n"
"gpr	r29	.32	124	0\n"
"gpr	r30	.32	128	0\n"
"gpr	r31	.32	132	0\n"
"gpr	cr	.32	136	0\n"
"gpr	xer	.32	140	0\n"
"gpr	lr	.32	144	0\n"
"gpr	ctr	.32	148	0\n"
"gpr	mq	.32	152	0\n"
"gpr	vrsave	.32	156	0\n"
);
```
test continues...

**Comment by unixfreaxjp** [comment]
test failed, condition: above patch was applied + --disable-debugger.
```asm
[...]

CC linux_debug.c
p/native/linux/linux_debug.c: In function 'linux_reg_profile':
p/native/linux/linux_debug.c:30:2: error: #error "Unsupported Linux CPU"
p/native/linux/linux_debug.c: In function 'linux_handle_signals':
p/native/linux/linux_debug.c:45:2: warning: #warning DO MORE RDEBUGREASON HERE [-Wcpp]
p/native/linux/linux_debug.c: In function 'print_fpu':
p/native/linux/linux_debug.c:293:2: warning: #warning not implemented for this platform [-Wcpp]
p/native/linux/linux_debug.c: In function 'linux_reg_read':
p/native/linux/linux_debug.c:378:3: warning: #warning not implemented for this platform [-Wcpp]
p/native/linux/linux_debug.c:385:4: error: unknown type name 'R_DEBUG_REG_T'
p/native/linux/linux_debug.c:298:7: warning: variable 'showfpu' set but not used [-Wunused-but-set-variable]
p/native/linux/linux_debug.c: In function 'linux_reg_write':
p/native/linux/linux_debug.c:427:43: error: 'struct user' has no member named 'u_debugreg'
p/native/linux/linux_debug.c:433:17: error: 'R_DEBUG_REG_T' undeclared (first use in this function)
p/native/linux/linux_debug.c:433:17: note: each undeclared identifier is reported only once for each function it appears in
p/native/linux/linux_debug.c: At top level:
p/native/linux/linux_debug.c:205:13: warning: 'print_fpu' defined but not used [-Wunused-function]
p/native/linux/linux_debug.c: In function 'linux_reg_profile':
p/native/linux/linux_debug.c:32:1: warning: control reaches end of non-void function [-Wreturn-type]
make[4]: *** [p/native/linux/linux_debug.o] Error 1
make[4]: *** Waiting for unfinished jobs....
make[3]: *** [foo] Error 2
make[2]: *** [debug] Error 2
make[2]: *** Waiting for unfinished jobs....
LD libr_asm.so
CC asm_propeller.c
CC propeller_disas.c
gcc: warning: /test/radare2/libr/../shlr/capstone/libcapstone.a: linker input file unused because linking not done
gcc: warning: /test/radare2/libr/../shlr/capstone/libcapstone.a: linker input file unused because linking not done
make[1]: *** [all] Error 2
make: *** [all] Error 2
Oops
~/radare2$
```
The p/native/linux/linux_debug.c needs more adjustment (more than the patch) for ppc CPU, either to be added in **linux_ppc.h** or in **linux_debug.c**

**Comment by unixfreaxjp** [comment]
shall this be considered as unsupported platform? @radare @crowell or to try a bit further?

**Comment by unixfreaxjp** [comment]
Modifying block codes of these errors:
```asm
[1] In function 'linux_reg_profile':
p/native/linux/linux_debug.c:30:2: error: #error "Unsupported Linux CPU"

[2] In function 'linux_reg_read':
p/native/linux/linux_debug.c:385:4: error: unknown type name 'R_DEBUG_REG_T'

[3] In function 'linux_reg_write':
p/native/linux/linux_debug.c:427:43: error: 'struct user' has no member named 'u_debugreg'
p/native/linux/linux_debug.c:433:17: error: 'R_DEBUG_REG_T' undeclared (first use in this function)
```

**Comment by radare** [comment]
i think __ppc__ is wrong. at least on OSX is __POWERPC__ i dont have a native linux on powerpc to test that thing but i pushed some blind commits in master. can you please retry building?

i think --disable-debugger is not properly working now. because all the native debugging capabilities should be avoided if that flag is passed to configure instead of failing here.

**Comment by unixfreaxjp** [comment]
> i think ppc is wrong. at least on OSX is POWERPC 
It explained:
```asm
$ cat ./libr/debug/p/native/linux/linux_debug.c | egrep '_ppc_|_POWERPC_'
#elif __ppc__
#elif __POWERPC__
#elif __POWERPC__
```
> i dont have a native linux on powerpc to test that thing but i pushed some blind commits in master. can you please retry building?

Retrying, pull all codes beforehand, with thank you.

>  think --disable-debugger is not properly working now. because all the native debugging capabilities should be avoided if that flag is passed to configure instead of failing here.

Noted, thanks very much for the important statement for the furture reference in similar cases avoiding debugging function build.

**Comment by unixfreaxjp** [comment]
Suggestion:
- ibr/.../p/native/linux/linux_debug.c need to be recoded to have skeleton that can support wider possible IoT (1) architecture and (2) cpu
- to consider fixing flag ./configure --disable-debugger for static function build only that can support the build in minimum IoT environment, by avoiding links to p/debug/native/linux/{`related codes'} during build.
- in linux_debug.c, to define flag powerpc or ppc (for ppc) and POWERPC for mac OSX should be cleared.
```asm
#elif __ppc__ || __powerpc__ || __POWERPC__ 
#include "reg/linux-ppc.h"
```
- maybe we should consider sparc too, I prepare the test environment as I'm typing this.
My point: It's critical issue that malware coder is already steps ahead than developer of tools to dissect them, why I push this matter.

**Comment by radare** [comment]
@zamaudio do you have a working qemu-ppc right?

@unixfreaxjp can you try submitting a pullreq?

**Comment by zamaudio** [comment]
I have one, but r2 doesnt compile there on my vm, a PR that works would be useful.

**Comment by unixfreaxjp** [comment]
> @unixfreaxjp can you try submitting a pullreq?

on it

**Comment by unixfreaxjp** [comment]
refer to https://github.com/radare/radare2/blob/master/libr/debug/p/native/linux/linux_debug.c
compiler seems went straight to line 377 -
```c
#else
	#warning not implemented for this platform
#endif
```
- and it was ignoring what is expected to be _ppc's_ l**inux struct fpregs** trap in (line: 350-373) by  !__ANDROID__, no? 
```c
#if !__ANDROID__
			struct user_fpxregs_struct fpxregs;
			ret1 = ptrace (PTRACE_GETFPXREGS, pid, NULL, &fpxregs);
			if (ret1 == 0) {
				if (showfpu) print_fpu ((void *)&fpxregs, ret1);
				if (sizeof(fpxregs) < size) size = sizeof(fpxregs);
				memcpy (buf, &fpxregs, size);
				return sizeof(fpxregs);
			} else {
				ret1 = ptrace (PTRACE_GETFPREGS, pid, NULL, &fpregs);
				if (showfpu) print_fpu ((void *)&fpregs, ret1);
				if (ret1 != 0) return false;
				if (sizeof(fpregs) < size) size = sizeof(fpregs);
				memcpy (buf, &fpregs, size);
				return sizeof(fpregs);
			}
#else
			ret1 = ptrace (PTRACE_GETFPREGS, pid, NULL, &fpregs);
			if (showfpu) print_fpu ((void *)&fpregs, 1);
			if (ret1 != 0) return false;
			if (sizeof(fpregs) < size) size = sizeof(fpregs);
			memcpy (buf, &fpregs, size);
			return sizeof(fpregs);
#endif // !__ANDROID__
```
**Below idea may help to fix these..**

I think by adding ppc's specific condition like mentioned area will fix the flow, since the **user struct fpregs** for ppc seems to be unrecognized, judging by the error messages _(confident: medium)_, this is what seems to be a problem in ppc compilation.

To be added (example) after line 349
```c
#elif __ppc__ || __powerpc__ || __POWERPC__
			ret1 = ptrace (PTRACE_GETFPREGS, pid, NULL, &fpregs);
			if (showfpu) print_fpu ((void *)&fpregs, 1);
			if (ret1 != 0) return false;
			if (sizeof(fpregs) < size) size = sizeof(fpregs);
			memcpy (buf, &fpregs, size);
			return sizeof(fpregs);
```
if so, we will need to change line 445 to out ptrace with , like: _(confident: low)_
```c
#elif __ppc__ || __powerpc__ || __POWERPC__
		int ret = ptrace (PTRACE_SETREGS, dbg->pid, &regs, NULL);
```
Why I think this will work, because please see line: 27-28. The grep for ppc flag architecture used is working in there: _(confident: high)_
```c
 #elif __ppc__ || __powerpc__ || __POWERPC__
#include "reg/linux-ppc.h"
```
So it is up to you to decide, first, please judge whether r2 will support this cpu, so we can hack malware & bad firmware in some routers with same arch like this. If this put too much work and need to focus to other development, cracking for this cpu's bad stuff will have to be done in r2 static mode, and that doesn't have to be done in this cpu, or do the ancient methods. I'll go either ways anyway.

Standing by :)

**Comment by radare** [comment]
Make a pull request please

**Comment by unixfreaxjp** [comment]
> Make a pull request please

Hello.
You mean me coding "__ powerpc __" compatibility in [p/native/linux/linux_debug.c](https://github.com/radare/radare2/blob/master/libr/debug/p/native/linux/linux_debug.c)?  If so I'll stick to 32bit only  (not adding any " __ powerpc64 __" stream since can't test that).
It isn't easy, on it. Now I'm forking all, designing/testing/coding after day work. 
Will pull request after my code tested. Cheers.

**Comment by radare** [comment]
You can do a pullreq via web just clicking in the pencil button, everything will be tested automatically

**Comment by unixfreaxjp** [comment]
[EDIT] forked+recoded PoC: [linux_debug.c](https://github.com/unixfreaxjp/radare2/blame/master/libr/debug/p/native/linux/linux_debug.c). Status: 1st & 2nd build failed many adjustment made. Not good, with this result I can not pullreq yet.
Below is last stage error:
```asm
p/native/linux/linux_debug.c: In function 'linux_reg_read':
p/native/linux/linux_debug.c:404:4: error: unknown type name 'R_DEBUG_REG_T'
p/native/linux/linux_debug.c: In function 'linux_reg_write':
p/native/linux/linux_debug.c:446:43: error: 'struct user' has no member named 'u_debugreg'
p/native/linux/linux_debug.c:452:17: error: 'R_DEBUG_REG_T' undeclared (first use in this function)
```

**Comment by radare** [comment]
r2 builds fine in powerpc now, but it fails to run because of endian problems. moving for 0.10.4

**Comment by radare** [comment]
please confirm that is fixed

**Comment by unixfreaxjp** [comment]
on testing mode now.. but during check last code `heads/master-0-g2f5ead1` before start tests this error came up, it seems like the latest modification is killing freebsd, and mostly all bsd will be affected in that same broken links https://github.com/radare/radare2/issues/5279
Please help to check and kindly let me know if you want to keep on testing arm/mips/ppc anyway

**Comment by unixfreaxjp** [comment]
I test built the committed code as per following detail:
```asm
[0x10001c0c]> !r2 -v
radare2 0.10.4-git 11827 @ linux-ppc-32 git.0.10.3-403-g2f5ead1
commit: 2f5ead1efd481af5864acad26fc7307d2555e8f3 build: 2016-07-04
[0x10001c0c]>
```
The Built test was okay, there is no error that can halt the compiling and linking process. However, [just like what happened in MIPS build ](https://github.com/radare/radare2/issues/4969#issuecomment-230260879)too, **the running test failed** with the below two symptoms:
1. Not recognizing the test architecture .. it is still fixable by firing `e asm.arch=ppc`.
2. Invalid data on `pd `and `px `which makes me stopping the further check, assuming same endian problem as per MIPS one.
The data of the test is in here: https://github.com/unixfreaxjp/dev/blob/master/r2ppc-6.md

**Comment by radare** [comment]
can you share that /bin/nc that is wrongly recognized? is this problem happening only on ppc? i guess thats an issue in the elf parser for bigendian. but at least it builds :D

**Comment by radare** [comment]
good news is that it compiles on both archs… and there are many endian issues in the elf parser, so that was expected. let’s see if i can have some time after release to fix them all.

thanks

**Comment by unixfreaxjp** [comment]
> good news is that it compiles on both archs… and there are many endian issues in the elf parser, so that was expected. let’s see if i can have some time after release to fix them all.

First of all. You did a great work in pinning the compilation process in PPC and MIPS. A tough task, so, congratulations again. :-)

> can you share that /bin/nc that is wrongly recognized? is this problem happening only on ppc? i guess thats an issue in the elf parser for bigendian. but at least it builds :D

I think it doesn't matter about the nc binary, the all binaries are read the same. In this below picture I tested with the "ls". And .. yes, it is the parser for big endian part. It's about time.
![](https://lh3.googleusercontent.com/-1HViyD3RW2Y/V3peyuzHsBI/AAAAAAAAViI/udZeXSuNiqM4vn4ZMIW01OwDL9a0ThhqACLcB/s720/3001.PNG)

No, thank you for achieving this far. PS: we have bigger issue for not compiling in freebsd x86 now.

**Comment by radare** [comment]
those bins load fine on little -endian. so thats a big-endian specific issue


thanks for the bins

**Comment by unixfreaxjp** [comment]
> those bins load fine on little -endian. so thats a big-endian specific issue

As per we discussed this before. Sooner or later the parser has to be adjusted to "BigEndian", yes it is a bit of work, but right now there's DVR, Webcam and TV starting to DDoS internet there, people will need r2 to run on them or its dev environment :) God speed.

**Comment by unixfreaxjp** [comment]
@crowell 

> (sparc should also fail because it doesn't have a reg profile in the debug plugin.)

What would you think about the usage of Linux/sparc in daily appliance so far, i.e. IoT?

**Comment by radare** [comment]
I'm closing this issue as long as compilation is fixed. Big endian issues will be fixed in the next release. still trying to setup a good-performing big-endian machine to work in

**Comment by unixfreaxjp** [comment]
> I'm closing this issue as long as compilation is fixed. Big endian issues will be fixed in the next release. 

copy that. new issue will be then?

**Comment by radare** [comment]
yep, open an issue to address the big-endian issues in the elf parser
